src/ast/ast_validations.c

```diff
@@ -1113,6 +1118,98 @@ static VISITOR_STRATEGY _Validate_CALL_Clause
 	return VISITOR_CONTINUE;
 }
 
+static void _AST_GetWithIdentifiers(
+	const cypher_astnode_t *node, 
+	rax *identifiers
+) {
+	if(!node) {
+		return;
+	}
+	ASSERT(identifiers != NULL);
+
+	cypher_astnode_type_t type = cypher_astnode_type(node);
+	if(type == CYPHER_AST_IDENTIFIER) {
+		const char *identifier = cypher_ast_identifier_get_name(node);
+		raxInsert(identifiers, (unsigned char *)identifier, strlen(identifier), NULL, NULL);
+		return;
+	}
+
+	uint child_count = cypher_astnode_nchildren(node);
+
+	// In case current node is of type projection inspect first child only
+	if(type == CYPHER_AST_PROJECTION) {
+		child_count = 1;
+	}
+
+	for(uint i = 0; i < child_count; i++) {
+		const cypher_astnode_t *child = cypher_astnode_get_child(node, i);
+		_AST_GetWithIdentifiers(child, identifiers);
+	}
+}
+
+static void _AST_GetWithReferences
+(
+	const cypher_astnode_t *node, 
+	rax *identifiers
+) {
+	if(!node) {
+		return;
+	}
+
+	if(cypher_astnode_type(node) != CYPHER_AST_WITH) {
+		return;
+	}
+	ASSERT(identifiers != NULL);
+
+	uint num_with_projections = cypher_ast_with_nprojections(node);
+	for(uint i = 0; i < num_with_projections; i ++) {
+		const cypher_astnode_t *child = cypher_ast_with_get_projection(node, i);
+		_AST_GetWithIdentifiers(child, identifiers);
+	}
+}
+
+static AST_Validation _Validate_Aliases_DefinedInWithClause(
+	const cypher_astnode_t *clause,
+	rax *defined_aliases
+) {
+	AST_Validation res = AST_VALID;
+	rax *with_aliases = raxNew();
+	rax *with_referred = raxNew();
+
+	raxIterator it;
+
+	// Get aliases identifiers in with clause
+	_AST_GetWithAliases(clause, with_aliases);
+
+	// Get referred identifiers in with clause, but don't include references of ORDER BY
+	_AST_GetWithReferences(clause, with_referred);
+	
+	_prepareIterateAll(with_aliases, &it);
+
+	// Check if every alias identifier that is referred was previously defined
+	while(raxNext(&it)) {
+		bool referred = false;
+		bool defined = false;
+		int len = it.key_len;
+		unsigned char *alias = it.key;
+		if(raxFind(with_referred, alias, len) != raxNotFound) {
+			referred = true;
+		}
+		if(raxFind(defined_aliases, alias, len) != raxNotFound) {
+			defined = true;
+		}
+		if(referred && !defined) {
+			ErrorCtx_SetError("%.*s not defined", len, alias);
+			res = AST_INVALID;
+			break;
+		}
+	}
```

Comment:
To avoid `if` branches, I suggest the following:
```sh
bool referred;
bool defined;
int len;
unsigned char *alias;
while(raxNext(&it)) {
		len = it.key_len;
		alias = it.key
		referred = (raxFind(with_referred, alias, len) != raxNotFound);
		defined = (raxFind(defined_aliases, alias, len) != raxNotFound);
		if(referred && !defined) {
			ErrorCtx_SetError("%.*s not defined", len, alias);
			res = AST_INVALID;
			break;
		}
}
```

Reply:
Change done. Thanks.